modules/core/upgrade.py

```diff
@@ -0,0 +1,551 @@
+
+
+def get_instance_ids_from_tag(tag_pattern):
+    try:
+        return set(invoke_shell("""
+        aws ec2 describe-instances --filter \
+            "Name=tag:Name,Values=*{}*" \
+            "Name=instance-state-name,Values=running" | \
+            jq --raw-output '.Reservations[].Instances[].InstanceId'
+        """.format(tag_pattern)).split())
+    except:
+        raise AssertionError(
+            'Cannot find instance IDs with tag pattern "{}"!'.format(tag_pattern))
+
+
+def get_instance_ip_addrs_from_tag(tag_pattern):
+    try:
+        return set(invoke_shell("""
+        aws ec2 describe-instances --filter \
+            "Name=tag:Name,Values=*{}*" \
+            "Name=instance-state-name,Values=running" | \
+            jq --raw-output '.Reservations[].Instances[].PrivateIpAddress'
+        """.format(tag_pattern)).split())
+    except:
+        raise AssertionError(
+            'Cannot get instance IP addresses with tag pattern "{}"!'.format(tag_pattern))
+
+
+def get_instance_ip_addr_from_id(id):
+    try:
+        return invoke_shell("""
+        aws ec2 describe-instances --filter \
+            "Name=instance-state-name,Values=running" | \
+            jq --raw-output '.Reservations[].Instances[] | 
+                select(.InstanceId == "{}") |
+                .PrivateIpAddress'
+        """.format(id)).strip()
+    except:
+        raise AssertionError(
+            'Cannot find instance IP address with Node ID: "{}"!'.format(id))
+
+
+def get_instance_ip_addrs_from_ids(ids):
+    return map(get_instance_ip_addr_from_id, ids)
+
+
+def get_new_instances_from_prev(prev_instances, curr_instances):
+    return curr_instances.difference(prev_instances)
+
+
+def kill_fn(aws_instance):
+    try:
+        invoke_shell("""
+        aws autoscaling terminate-instance-in-auto-scaling-group \
+            --no-should-decrement-desired-capacity \
+            --instance-id {}
+        """.format(aws_instance))
+    except:
+        raise AssertionError('Unable to terminate AWS instance "{}" from its ASG!'
+                             .format(aws_instance))
+
+
+def try_until_timeout(check_fn, prev_aws_instances, killing_idx, check_interval, timeout):
+    elapsed_time = timedelta(seconds=0)
+
+    while True:
+        if check_fn(prev_aws_instances, killing_idx):
+            return True
+
+        if elapsed_time >= timeout:
+            return False
+
+        wait_secs = check_interval.total_seconds()
+        print('> Waiting for {}s ({}s elapsed)'.format(
+            wait_secs, elapsed_time.total_seconds()))
+        time.sleep(wait_secs)
+        elapsed_time += check_interval
+
+
+def check_service_up(prev_aws_instances, killing_idx, n, tag_pattern, get_service_nodes_fn):
+    # We assume that prev_aws_instances contain the same number of entries as
+    # the original
+    instance_count = len(prev_aws_instances)
+    curr_aws_instances = get_instance_ids_from_tag(tag_pattern)
```

Comment:
Would probably be more helpful if there's a higher level check to validate the cluster. For e.g. get the number of nomad servers in the _cluster_ by querying a well-known endpoint

Reply:
I realised it's done here: https://github.com/GovTechSG/terraform-modules/pull/213/files#diff-106171020776ac2d38fc8b0f3c8c960cR227